### ISSUE TITLE
Proposal: split bitcoin into multi-crate workspace

### ISSUE THREAD

**Issue by apoelstra** [rust-bitcoin/rust-bitcoin#550]
I propose we split some parts of rust-bitcoin into multiple crates:
* `bitcoin-encoding` which has all the string encodings that bitcoin uses: base58, base58check, bech32, bech32m, base64, hex
  * this would collapse `base64_compat`, `bech32-bitcoin`, `byteorder` all into one dep that we control. maybe also `bech32`.
  * would also let us write a unified interface to all these
  * an allmost-natural place for us to stick the endianness code :P
  * potentially could be a dep of rust-secp, letting us encode keys and base64 signatures the way users expect
  * lets us deal with the unique no_std difficulties around string encoding in one place
* `bitcoin-primitives` which defines block/transaction/script, and the p2p messages, and the consensus::Encodable trait
  * lets users parse/serialize bitcoin stuff without needing rust-secp and the rest of the massive dep tree. see e.g. #527 
  * similar benefits to the above wrt no_std, lets us deal with `Vec` encoding and use of `io::Write` in one place
* `bitcoin-keys` which deals with private/public keys, addresses, bech32, taproot, as well as computing sighashes. I think this is the only one that'd require a rust-secp dependency
* `bitcoin-psbt`

I think this covers all the contents of the library. Then the main "bitcoin" crate would simply have optional deps on all of the above and re-export everything from them.

**Comment by apoelstra** [comment]
I think that splitting out `bitcoin-encoding` at least is a good idea and that we should pursue it, maybe for rust-bitcoin 0.27.

The other crates I could take or leave.

**Comment by Kixunil** [comment]
I'd like to understand more about how would you like to split **types**. I could imagine some of them in both `bitcoin-primitives` and `bitcoin-encoding`.

E.g. `Txid` has stringly encoding but is also used in `OutPoint`. Or maybe even more natural hypothetical `Amount(u64)` type in the future. Addresses have stringly implementation but also cryptographic operations.

I think it'd be clean to have them in a separate crate but `primitives` sounds more natural.

So perhaps:
* `bitcoin-primitives` - types shared by at least two crates below; also known as "entities" in Clean architecture
* `bitcoin-str` - implementation of stringly encoding
* `bitcoin-consensus-codec` - implementation of consensus encode/decode
* `bitcoin-keys` (or perhaps `bitcoin-crypto`) - implementations of cryptographic functions
* `bitcoin-psbt`

**Comment by dr-orlovsky** [comment]
For rust-psbt I already had a work in progress, which I can update for the latest developments and migrate it as a crate into rust-bitcoin with a PR: https://github.com/LNP-BP/rust-psbt

**Comment by apoelstra** [comment]
@Kixunil so, I think basically everything would have to depend on `bitcoin-str` (or `bitcoin-encoding`, though I like your name better). Then the separation of types would be

* `Block`, `Transaction`, `Script`, `Address`, `Txid` (and all other hashtypes) in `bitcoin-primitives`, and they would impl `fmt::Display` etc using `bitcoin-str`
* `PrivateKey`, `PublicKey` would be in `bitcoin-keys`, which would depend on `bitcoin-primitives`, and the conversion methods from keys to addresses would live there. Ditto for bip32 types
* Then I *think* `bitcoin-psbt` can depend on -primitives and -str but it shouldn't need to depend on -keys.

**Comment by apoelstra** [comment]
I'm not sure there's a big benefit to splitting `bitcoin-consensus-codec` out of `bitcoin-primitives`. We don't intend for non-bitcoin types to use the consensus encoding and it's hard to use any of the primitives without it.

**Comment by Kixunil** [comment]
I think `bitcoin-str` could also be optional (but maybe it's reasonably small?). I wouldn't be surprised if some applications didn't use it if they operated on binary data only. Although maybe logging is too widespread to not use it.

I could imagine an application that does e.g. accounting and loads something like `address, amount, description` CSV and wants to decode `address` as a sanity check, so such application wouldn't need consensus encoding-decoding. Perhaps not convincing example to bother but maybe there's a better one.

**Comment by apoelstra** [comment]
Many things in the Core RPC are hex-serialized versions of binary-encoded data. (In my usage, almost 100% of my RPC interaction is of this form.) So I don't think you could work with the RPC without having the consensus encoding.

**Comment by thomaseizinger** [comment]
> I think that splitting out `bitcoin-encoding` at least is a good idea and that we should pursue it, maybe for rust-bitcoin 0.27.
> 
> The other crates I could take or leave.

Agree with this.
Splitting out the encoding functionality is useful because it means `rust-secp256k1` can depend on it without a circular dependency.

For all the remaining functionality, I think we should consider feature-flagging it first. Especially if the main `bitcoin` crate is meant to re-export all of the potential crates anyway.

I like that `bitcoin` is almost a one-stop-shop for working with bitcoin in Rust. Personally, I would even like to have all the RPC types in `bitcoin` as well (feature-flagged to avoid bloat if not needed). Currently, the bitcoin-rpc-json crate often lags behind a version for quite some time which prevented us more than once from updating to the latest version of `bitcoin`.

A downside of feature-flags is that a minor version bump (we are still < 1.x) affects all types where as with separate crates, those could be versioned independently. At the same time, if more types (like `bitcoin-json-rpc`) were to reside in `bitcoin`, these version bumps would be less painful because you don't have to update several dependencies at the same time.

**Comment by dr-orlovsky** [comment]
The most of projects using `rust-bitcoin` do not require network message part, so it either may be feature-flagged (and not being default) or moved to a separate crate.

**Comment by sgeisler** [comment]
+1 for encoding. I'm not sure if the rest is worth the hassle. What's the concrete goal here? Who actually _wants_ to use some parts without pulling in the rest? LTO should get rid of unused code anyway, doesn't it? I totally support breaking things up to reuse code, but without concrete objectives I'd like to avoid the churn. I also don't really fancy update-cascades when changing some type deep down in the dependency tree.

Regarding bech32 we should probably ask @clarkmoody, since it's already used outside our ecosystem I'd rather not discontinue it in its current form.

**Comment by clarkmoody** [comment]
A `bitcoin-encoding` crate could easily re-export `bech32`, or wrap the operations in a common interface.

There will be necessary changes made to that crate with the upcoming Bech32m encoding. This would be a good opportunity to either deprecate the crate and move everything to `bitcoin-encoding` or to revisit the crate and make necessary updates and fixes.

Either way is good for me.

**Comment by Kixunil** [comment]
> LTO should get rid of unused code anyway, doesn't it?

It needlessly increases compile times, LTO even more. That being said, it'd be nice if we could measure it without doing the work of splitting up.

I'd also really like to see `secp256k1` separated as C FFI can occasionally be hell so it'd be great if crates that don't need it didn't experience its problems.

> I also don't really fancy update-cascades when changing some type deep down in the dependency tree.

A disincentive to breaking public APIs may be a good thing. Would love to see `rust-bitcoin` stabilized eventually. Splitting up can actually help that - it allows putting stable APIs into one crate and unstable into another. My guess is that stabilizing some primitives could be done soon.

**Comment by Kixunil** [comment]
Just remembered that I was playing with `-Z timings` yesterday and tried it on `electrs` looking at its output, `bitcoin` takes 26 seconds to compile, half of that is codegen. That is worth optimizing a bit IMO.

**Comment by JeremyRubin** [comment]
I like splitting things up. I also generally prefer more crates to fewer. Even things like the encoding trait I think could/should be defined in their own crate (& optionally re-exported for convenience) so they can be pulled in with minimal dependencies in certain contexts. (The rules for definition mean the trait impls for external types have to live with the trait itself, the rest of the impls for our defined types can live where they are defined). For the encodings I also like splitting it into sub-crates for each type and making one big package that re-exports all the modules so they can be pulled in a-la-carte if needed.

These don't need to be sep repos, just sep workspaces in one repo (altho they could be sep eventually). That helps make it so the crates feel more like modules with stricter boundaries.

**Comment by dr-orlovsky** [comment]
@apoelstra I agree with @Kixunil on consensus encoding to be separate and will add an argument that many projects will either need things like `bech32*`(wallet types) or `consensus_encode` (working with blockchain directly or P2P network).

So I propose to have a tree of crates like
- `bitcoin-encodings` (group crate reexporting stuff from below)
  + `bech32`
  + `bech32m`
  + `base58` (seems you forgot about it)
  + `base64_compat` as `base64`
  + `byteorder` 
  + `consensus_encode`

Also, I think that stuff like BIP32 deserves separate crate, alongside PSBT. I have developed a lot of BIP32-related code which I can propose as a candidate on this role.

Another point: hashtypes (like Txid) may deserve to be a part of `bitcoin_hashes`, since they are frequently used when other primitives (like `Transaction`, `Block`) are not required – and you can't have them w/o hashes themselves.

Think not all those crates can be a part of the same repo, `rust-bitcoin`; stuff like `psbt` and `bip32` may be separate, or be moved into a abandoned `bitcoin-wallet` crate.

**Comment by Kixunil** [comment]
@dr-orlovsky 

> Also, I think that stuff like BIP32 deserves separate crate

Yes, that makes a lot of sense

> since they are frequently used when other primitives (like `Transaction`, `Block`) are not required

I don't remember ever seeing such software, do you have an example?

Into which crate would you put addresses?

**Comment by dr-orlovsky** [comment]
> I don't remember ever seeing such software, do you have an example?

Database-stored indexes. Software reading them (APIs etc) does not need to know anything about transaction or blocks tructure, but must operate hash-based types.

> Into which crate would you put addresses?

I think they are more appropriate in wallet type library; but ok if they will be kept anywhere else.

**Comment by Kixunil** [comment]
Ah, I see, so essentially block explorers and maybe other analysis tools use hashes, that sounds reasonable.
I think that addresses are similar in this regard. They are all primitives/entities in my mind.

**Comment by dr-orlovsky** [comment]
>  They are all primitives/entities in my mind.

They are, but with a very big difference: addresses and keys require secp256k1 lib as a dependency (with FFIs), while the rest of primitives do not.

**Comment by apoelstra** [comment]
You can't even compute txids without `consensus_encode`. I'm really having trouble imagining what kind of bitcoin software would not need to encode transactions ever.

**Comment by apoelstra** [comment]
I would also not like to split `bitcoin_encoding` into sub-crates. All of these encodings are 10-20 lines of "core" code plus boilerplate that we could probably unify between them. Splitting them up might make this unification harder, and increases the apparent size of the dep tree for almost no gain.

**Comment by JeremyRubin** [comment]
I'd also point out that we can guarantee that the txid type serializes and deserializes (even in a bitwise copy via repr transparent) identically to the inner hash, so I think it's reasonable that if you want just a blob store you don't really need the txid type itself.

OTOH the way we currently create the TXID type is based on a macro defined in the parent crate, so it doesn't have access to the local crate's  stuff except for purpose of deriving trait Encodable.

So if Txid were moved to bitcoin-hashes, it would impose that bitcoin_hashes should depend on the encoding trait crate (or vice versa?) in order to be able to impl Encodable.

**Comment by apoelstra** [comment]
@JeremyRubin I'm not fully sure I understand. We could have the TXID type in bitcoin_hashes, but the code for constructing it from a transaction would live in bitcoin-primitives, with no conflict there.

For what it's worth a txid is different from a normal hash in that it's serialized backward, so there is some value in having a special-purpose type.

**Comment by Kixunil** [comment]
> addresses and keys require secp256k1 lib as a dependency

Not entirely, you can encode/decode/compare/hash/... them without any cryptography. Usable in indexers. Although maybe not needed to separate it that way today. `electrs` doesn't need `secp256k1` but it works with script hashes instead of raw addresses and I don't know of any other widely used indexer that works differently.

> if you want just a blob store you don't really need the txid type itself.

I like newtypes... :)

**Comment by JeremyRubin** [comment]
> @JeremyRubin I'm not fully sure I understand. We could have the TXID type in bitcoin_hashes, but the code for constructing it from a transaction would live in bitcoin-primitives, with no conflict there.
@apoelstra 
currently the Txid is defined in the same crate as the Encodable/Decodable trait, so it's possible to impl Encodable/Decodable. If Txid were moved to bitcoin_hashes, and encodable were to stay inside of rust_bitcoin, that would also be ok. But if Encodable moves to its own crate, and Txid moves to bitcoin_hashes, then you need to make either encodable's crate depend on bitcoin_hashes, or bitcoin_hashes' crate depend on encodable in order to implement the trait.

**Comment by apoelstra** [comment]
Ah! Yes, if we were to move the consensus-encoding trait outside of rust-bitcoin, it'd have to be into bitcoin-primitives, and then every single plain-old-data type that appear in transactions/blocks, including `Txid`, would have to follow.

**Comment by dr-orlovsky** [comment]
I tried to summarize possible crates and their dependencies in a chart according to the discussion above sticking to the most commonly agreed part. Pls propose what I miss here

![image](https://user-images.githubusercontent.com/372034/105615178-e9144d00-5dce-11eb-9e6c-3c2f44bfbf63.png)

**Comment by dr-orlovsky** [comment]
Or, as an alternative - trying to fit validation with`bitcoinconsensus`

![image](https://user-images.githubusercontent.com/372034/105615395-9f2c6680-5dd0-11eb-8c6d-35cb589550de.png)

**Comment by Kixunil** [comment]
Nice, could you draw arrows instead of lines so we know how things depend on each-other?

**Comment by dr-orlovsky** [comment]
@Kixunil for sure, but the rule is simple: the thing on top depends on a bottom one

**Comment by Kixunil** [comment]
That makes sense. I'm just not sure about relationship between primitives and encoding - it feels more correct to reverse it. (I believe in Clean Architecture but willing to reconsider if there's a good criticism of it.)

**Comment by dr-orlovsky** [comment]
@Kixunil sIs there any need in bitcoin primitives w/o ability to encode/decode them? It  will be really odd when you need just a `Txid` to be read/written to have two dependencies instead of one, on `bitcoin-primitives` and `bitcoin-encoding`.

**Comment by JeremyRubin** [comment]
Excellent work @dr-orlovsky! 

No it's correct to have encodings have minimal dependencies IMO. Because it's very feasible that you would want Bitcoin's encoding stuff, but *not* want rust-bitcoin's concept of Bitcoin's primitives. But you would never not want rust-bitcoin's encoding stuff if you are using rust-bitcoin's primitives.

**Comment by Kixunil** [comment]
@dr-orlovsky if my understanding is correct, `bitcoin_encoding` contains only stringly encoding. An application that works with raw binary data only doesn't need stringly encoding.
A serious Bitcoin application without any of these seems inconceivable:

* Amount
* Address
* Script
* Txid

So according to Clean Architecture they are entities and should be base of all other things.

**Comment by dr-orlovsky** [comment]
That's why I originally proposed to split `consensus-encoding` (which is the most important part of the encodings) from the rest (which is already in separate crates), namely `base58`, `base64(-compat)`, `bech32(m)` (you call them "stringly encodings").

In fact, consensus encoding is related to the system level and important for all software (since any bitcoin-related soft will need to read/write binary txes), while stringly encodings are only for wallet-related stuff.

**Comment by Kixunil** [comment]
So `bitcoin_encoding` in your drawing means consensus encoding? That'd make sense then but I wonder if splitting them is useful if there's pretty much zero chance anyone will use one without another.

**Comment by dr-orlovsky** [comment]
@Kixunil wallets will use both, which is the main use case :) And even blockchain explorers will. Anything interfacing humans. But I agree on separating binary encodings from the other. It's just pity that `bitcoin-hashes` can't be done without "stringly" `hex`,  since you would not be able to derive `FromStr` on foreign type above bitcoin-hashes, and everything else is dependent on hashes. The same applies to addresses, leading to `bech32(m)` and `base58`, so we end up with everything mixed once again...

**Comment by sgeisler** [comment]
I mean we can always feature-gate our stringly encodings behind a default-feature. I don't think we need a parent-crate for stringly encodings though, just pull in whatever you need to impl stringly encoding for the types you are defining. This would be similar to our handling of serde.

**Comment by Kixunil** [comment]
That's true, even for debugging printing strings is much better. In that case it's probably better to just join them, I don't see a significant advantage of splitting. We could revisit it later anyway.

Sadly, Rust orphan rules are a major roadblock when doing Clean Architecture. I had an idea for proposal on solving that but it's quite exotic, so I'm not eager to pursue it assuming low likelihood of accepting. (Short version is to allow explicitly replacing arbitrary trait impl in binary/dylib/staticlib crates and to also allow special impl crates that can only be used by binary/dylib/staticlib crates.)

**Comment by dr-orlovsky** [comment]
@sgeisler where bitcoin consensus encoding will reside? And if we need hex in secp256k1 than we have to separate hex crate from the rest.

Also, just a note: the difference between crate and repo in my opinion is the fact that the repo is a review/access control thing; while the crate is the "use scenario" thing. So merging consensus-critical & highly conservative code with non-critical in the same repo probably is not good. Nevertheless all encodings seems to be conservative, so this does not change anything (but bip32/psbt may be kept as a separate repo)

**Comment by sgeisler** [comment]
@dr-orlovsky I think consensus encoding traits might be an extra crate below bitcoin-primitives, but only if that's needed somewhere (e.g. some generic code that accepts any consesus encodable data but doesn't need primitives, which seems unlikely but possible). Otherwise I'd just put them into the primitives crate, where most of the structs need an impl of these anyway. The implementations would always reside where types are declared, potentially feature-gated (like with serde, which seems to work well).

Thinking about it we might want the consensus crate to depend on primitives to implement the conversion to C types and to present a more rustic interface.

**Comment by apoelstra** [comment]
rust-secp has its own copy of the hex encoding, and has no dependencies of its own.

I'm glad that @Kixunil pointed out how difficult Rust's orphan rules make Clean Architecture.

It is also worth considering the maintenance complexity of having many crates, and the difficulty in auditing large dependency trees. If rust-bitcoin had a dep tree the size the picture @sgeisler posted, I would fork the project and recombine everything to avoid having to deal with that noise. Many people in this conversation have actively avoided e.g. tokio for this reason.

I have no idea what @JeremyRubin means by "rust-bitcoin's concept of Bitcoin's primitives" or why you would not want it. Except for the introduction of segwit, these types have never changed, and they are plain-old-data types which directly correspond to objects encoded in the blockchain and peer-to-peer layer.

**Comment by sgeisler** [comment]
@apoelstra I don't think anyone will do this just to see you fork off :laughing: But wouldn't the dependency tree only _look_ big? It isn't really if the set of people with merge access is the same/they even reside in the same repo. The main problem appear to be outside dependencies not under our strict control. The main annoyance is getting all the versions right if there are no re-exports or you want to activate features in dependencies. I generally saw the rust ecosystem moving to more focused, smaller crates, I didn't experience it as something too horrible so far (especially if a bunch of crates comes from the same repo anyway).

**Comment by Kixunil** [comment]
Frankly people who judge a project purely on *count* of *formal* dependencies sound to me similar to big blockers. They focus on one magic number and want to change it without thinking about broader context. (Disclaimer: all people in this thread sound reasonable to me, I'm just pointing out theoretical extremes.)

Being afraid of *count* of "Compiling" lines printed by the compiler is not helpful. I remember avoiding `rust-bitcoin` in the past for some things because it seemed bloated to me by looking at all the stuff there that I don't need.

That being said, the nano-library approach promoted by NPM ecosystem is other extreme, so I don't want to go that direction. (I could tell you stories how bad it is...) Especially if there's a good argument about something being useless without other thing. Drawing more granular picture could help with reasoning but we need to draw lines around boxes in such picture.

I guess small, focused crates can reach 1.0 sooner which could stabilize whole thing and "getting versions right" is thus resolved.

**Comment by apoelstra** [comment]
I think I need to try using crev (it's long overdue) before commenting further on the size of the dep tree. I have a pretty negative view of the tooling that we have for dealing with this, but this may be my fault for not trying the things we have available :).

**Comment by NCrashed** [comment]
Hi, I am highly interested in `bitcoin::consensus::encode` module as separate library. It is great for hand crafted network protocols with endianess normalization and I wasn't able to find any already existing serialization libraries for such task.

**Comment by NCrashed** [comment]
As temporal workaround I made https://github.com/NCrashed/consensus-encode that is simple copy-paste and cleanup of `bitcoin::consensus::encode` module.

**Comment by TheBlueMatt** [comment]
The nice thing about Rust is its quite good at stripping unused code from final binaries, even if there's lots in a single crate...LTO is great (at least as long as you aren't linking in C symbols), so I'm a little dubious of separate crates for just that reason. That said I'm not against splitting the crate into several crates, though if we end up with multiple repos it gets much more complicated.

**Comment by Kixunil** [comment]
Sadly LTO doesn't decrease code review and compile times. :(

**Comment by TheBlueMatt** [comment]
Sure, I was just saying that splitting a crate up just to have smaller crates I really don't buy - compile time is once per version since rust is also pretty decent at caching build output. As for code review, I'm not sure how this will change? If we split the repo (which I don't think is being discussed here), if anything, code review will not happen slower, and if we don't I don't think it will be impacted.

**Comment by Kixunil** [comment]
I mean, if someone is adding a dependency and needs to review it, then having to read smaller code is helpful.

**Comment by RCasatta** [comment]
I think I really would love a `bitcoin-primitives` crate, I don't think the reason has been mentioned here. At the moment when downstream libraries like bdk need to update bitcoin version, they basically need to wait for other libs to update first (rust-electrum-client, rust-bitcoincore-rpc, rust-miniscript, bitcoinconsensus...) because struct like Txid is shared around.

bitcoin-primitives are really stable in comparison to the rest and would desyncronize the upgrade process

**Comment by Kixunil** [comment]
@RCasatta that's exactly what I meant by https://github.com/rust-bitcoin/rust-bitcoin/issues/550#issuecomment-760312696 :)

**Comment by RCasatta** [comment]
> @RCasatta that's exactly what I meant by [#550 (comment)](https://github.com/rust-bitcoin/rust-bitcoin/issues/550#issuecomment-760312696) :)

very long thread! 😅

**Comment by Kixunil** [comment]
Another datapoint: it would make more sense to only depend on bitcoin primitives from `bip21` and `ln-types` crates, not on the whole `bitcoin`.

**Comment by tcharding** [comment]
> A `bitcoin-encoding` crate could easily re-export `bech32`, or wrap the operations in a common interface.
> 
> There will be necessary changes made to that crate with the upcoming Bech32m encoding. This would be a good opportunity to either deprecate the crate and move everything to `bitcoin-encoding` or to revisit the crate and make necessary updates and fixes.
> 
> Either way is good for me.

Hi @clarkmoody! In light of this comment you made and the fact that the bech32 repo seems a bit dormant I might go ahead and merge the code from `bech32` into the new `bitcoin-str` crate create [here](https://github.com/rust-bitcoin/rust-bitcoin/pull/751). I just wanted to check in with you since the comment is quite old and just to make absolutely sure I'm not treading on you toes.

Thanks

**Comment by clarkmoody** [comment]
> Hi @clarkmoody! In light of this comment you made and the fact that the bech32 repo seems a bit dormant I might go ahead and merge the code from `bech32` into the new `bitcoin-str` crate being created [here](https://github.com/rust-bitcoin/rust-bitcoin/pull/751). I just wanted to check in with you since the comment is quite old and just to make absolutely sure I'm not treading on you toes.

Hey Tobin. Dormancy is not necessarily a bad thing, since the lib is very simple and the spec hasn't changed. 😉 We made the addition of Bech32m after my comment was posted.

There are a [good number of projects](https://crates.io/crates/bech32/reverse_dependencies) that rely on `bech32` externally, since it does provide single-dependency codec support for Bech32. I wouldn't want to do anything to rug pull those other projects.

If other contributors think it's better to roll the library into this repo, then we can certainly make a copy (and maintain both implementations going forward), but so far it doesn't look like that's the case.

**Comment by Kixunil** [comment]
I certainly don't believe forking `bech32` would be a good idea. Especial since it's in this org.

I'd love some polishing effort and make it 1.0 though. I think it could be done quite soon given it doesn't change much.

**Comment by apoelstra** [comment]
I would also be willing to help. Since this thread was last active, I have learned all the relevant coding theory to understand bech32. My reason for avoiding it in the past is that I didn't really understand the math, and there's nothing but math in that crate.

cc @clarkmoody would you be willing to add some extra maintainers to `bech32`?

I also agree with the sentiment that we shouldn't fork, it, just re-export it from `bitcoin-str`, since it's very simple and compact and useful beyond just Bitcoin.

**Comment by tcharding** [comment]
Re-export it is! Thanks.

**Comment by clarkmoody** [comment]
> cc @clarkmoody would you be willing to add some extra maintainers to bech32?

@apoelstra You should be a maintainer already.

**Comment by tcharding** [comment]
> I don't think we need a parent-crate for stringly encodings

@sgeisler the `endian` module needs a home, it is used by `base58` module. In your diagram we would either need _another_ crate or put it in with `base58`.

**Comment by apoelstra** [comment]
>@apoelstra You should be a maintainer already.

Oops, yeah, that makes sense as part of the transition to the rust-bitcoin org. I had totally forgotten! Will take a look at some of Tobin's PRs then.

**Comment by dr-orlovsky** [comment]
@Kixunil why this can't be a part of `bitcoin-primitives`?

**Comment by Kixunil** [comment]
Mainly because consensus encoding traits are currently poison for `no_std` and less things make it easier to decide on design and commit to keeping it stable.

**Comment by Kixunil** [comment]
Although the feature for enabling consensus encoding will have to be unsstable. I think it's OK.

**Comment by dr-orlovsky** [comment]
@afilini I am afraid traits would not be that efficient since you will end up with having `Box<dyn Trait>` everywhere upstream and end up with `type A cant't be maid into object` quite soon all around.

**Comment by Kixunil** [comment]
@afilini Intuitively this looks big and complicated and the upside seems questionable. I doubt you can cover many chains in the wild because of their significant differences and potential future changes. But maybe I'm wrong and there are some that are worth it. I'd however suggest to start abstracting from higher-level. Also if it's true these things are abstractable it should be possible to do in an external crate. Also reminds me of `num-traits` crate which I think we could and probably should support here.

Anyway, I wouldn't obstruct well-designed PR which demonstrates it's possible and practical.

**Comment by afilini** [comment]
> @afilini I am afraid traits would not be that efficient since you will end up with having Box<dyn Trait> everywhere upstream and end up with type A cant't be maid into object quite soon all around.

I don't think that's always true, you don't necessarily have to use dynamic dispatching. But it's true that you would have a lot more generics around probably. And also more methods in general because you wouldn't be able to access the struct members directly. But that would only be in the "primitives" crate, because everything else is by definition bitcoin-specific, so things like the sighash computation would only be implemented on `bitcoin::Transaction`, not on a generic `T: TransactionTrait` type.

> Also if it's true these things are abstractable it should be possible to do in an external crate.

That's a really good point, this could stay in another crate alongside `bitcoin-primitives` or whatever and re-expose its API via traits

**Comment by apoelstra** [comment]
I could be convinced that abstracting over the `Amount` field could be worthwhile.

**Comment by tcharding** [comment]
As we work on this I think managing crate features down our stack is going to come up. Flagging here a closed PR I did to start some discussion on this topic.

https://github.com/rust-bitcoin/rust-secp256k1/pull/408#pullrequestreview-888794047

**Comment by tcharding** [comment]
Relevant crate smashing discussion: https://github.com/rust-bitcoin/rust-bitcoin/discussions/921#discussioncomment-2487586

**Comment by tcharding** [comment]
Relevant comment: https://github.com/rust-bitcoin/rust-bitcoin/pull/1041#issuecomment-1157700244

**Comment by tcharding** [comment]
Relevant comment: https://github.com/rust-bitcoin/rust-bitcoin/pull/1062#issuecomment-1164449748

**Comment by tcharding** [comment]
> bitcoin_hashes is a separate crate so that I can do signature stuff without dragging in all of rust-bitcoin. [0]

Is there any reason `bitcoin_hashes` is in a separate repo? During crate smashing should we be thinking about repo separation? Is there any benefit in moving things in/out of `rust-bitcoin`?


[0] : bitcoin_hashes is a separate crate so that I can do signature stuff without dragging in all of rust-bitcoin.

**Comment by Kixunil** [comment]
I did think about it before and there would be benefits in both during initial smashing phase: we could test new hex stuff depending on it from `bitcoin_hashes` if they were a single repo but on the other hand, we wouldn't have to wait for PRs to merge before doing other work if they were separate. I'm not sure what's better.

**Comment by apoelstra** [comment]
@tcharding just historical reasons. I don't think workspaces existed when I started this project.

As we've noticed, having it in a separate repo causes some logistical challenges ... inconsistent CI scripts which aren't well-maintained, different sets of maintainers (and unlike rust-secp this is not a deliberate policy). My vote is to pull it into this repo.

**Comment by tcharding** [comment]
I had a moment of clarity when I woke up this morning, thinking about crate smashing. What if we inline into `base58` crate the required sha256d logic and then the circular dependency disappears (hex is only used in testing so some hex parsing code can be added to tests trivially too). Seems by now the only improvements to hashing are likely going to be optimizations and since base58 isn't used that much it seems reasonable to leave it without optimisations.

**Comment by apoelstra** [comment]
Hmm, yeah, I could get behind that. Then `bitcoin-str` could continue to have no dependencies, and also include base58.

**Comment by tcharding** [comment]
That wasn't too hard :), please see #1294.

**Comment by Kixunil** [comment]
Hmm, doesn't feel right. Wouldn't it be better to just not have `bitcoin-str` and only depend on specific things we require?

* `bitcoin_hashes` depend on `bitcoin-internals` (and if we **really** want to separate hex, have `bitcoin-hex`; but `internals` will still be needed for error handling)
* `base58` depends on `bitcoin_hashes`
* `address` (or whatever the crate it'll be in) depends on `base58`
* `PrivateKey` depends on `base58` and `secp256k1`

**Comment by apoelstra** [comment]
It really kills me to have a whole extra dependency just for `hex`, which probably would then need to also depend on some other crate to get the `ByteEncoder` trait or whatever, when we have several other formats which are basically the same and are all very small amounts of code.

If `cargo tree` were less idiotic maybe I would change my mind but it isn't, and I'd rather not spam up people's dependency trees by exporting our entire module system into it.

**Comment by Kixunil** [comment]
Yep, I don't want hex to be a separate crate, I want it in internals where there are other important things we will use anyway.

**Comment by apoelstra** [comment]
Would the `ByteEncoder` traits also be in -internals?

That sounds fine to me, except for the "we will never stabilize this" warning that we have on -internals. Maybe eventually we could have a whitelist of modules in -internals that we consider "1.0"?

**Comment by Kixunil** [comment]
I intended to have it in the same crate which defines consensus encoding. However if there is any reason to have it in internals then I suggest reexporting it in all crates it's used in and declare that whatever is re-exported like this is considered stable when accessed through a re-exported path (but accesses through `bitcoin-internals` are still unstable).

**Comment by apoelstra** [comment]
Sorry, I think I misspoke -- I meant `hex::BufEncoder`, not `ByteEncoder`.

**Comment by Kixunil** [comment]
Why would we need to expose `hex::BufEncoder` to the consumers as stable rather than providing APIs via `Display` & co?

**Comment by apoelstra** [comment]
`Display` is not a useful API because it requires having a `fmt::Formatter` which is opaque, unconstructable and of limited use. `Write` is not useful because it requires std.

I'm not aware of other generic "encode things into something" APIs, so I think we ought to expose *something*.

Maybe not, I guess there's no particular reason to be generic over hex/base64/bech32/etc.

**Comment by Kixunil** [comment]
If you're writing hex anyway just have `fmt::Write` which is available in `core`. Sure, the error type sucks but it's at least object-safe. But I'm not sure what the requirements are, so the question is where exactly does this even matter? Which project needs to hex-encode some `bitcoin` type and can not go through `Display`?

**Comment by apoelstra** [comment]
If I have an existing buffer that I want to write hex into can I do that with `fmt::Write`?

**Comment by RCasatta** [comment]
You can use `write!` https://doc.rust-lang.org/std/macro.write.html

**Comment by apoelstra** [comment]
Okay, I'm happy with that.

**Comment by Kixunil** [comment]
Well, `write!()` requires `std` as it internally goes through `io` when used with `&mut [u8]` but anyone needing `core` could implement a wrapper around `&mut [u8]` to write strings into.

**Comment by RCasatta** [comment]
I implemented the wrapper in the past for my embedded toy project, and later found and used `heapless::String` (use const-generics)

**Comment by jeandudey** [comment]
FWIW I can donate `bitcoin-amount` crate that is on crates.io for the `Amount` types.

**Comment by apoelstra** [comment]
Neat! Could you open a discussion on the Discussions tab about this?

It can be a short one-line topic, I don't wanna make you do a bunch of work, but it will give us one place to discuss the merits of this.

**Comment by tcharding** [comment]
At some stage it would probably be useful to go over this thread and pull out all the relevant ideas because a fair bit has changed.

**Comment by apoelstra** [comment]
Yeah. Let's say that we can close it with a file (or files) in `docs/` that copy down all our current plans.
